src/framework/audio/internal/sequencer.cpp

```diff
@@ -0,0 +1,291 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "sequencer.h"
+#include "internal/midiplayer.h"
+#include "internal/audioplayer.h"
+
+using namespace mu::audio;
+
+Sequencer::Sequencer()
+{
+    m_clock = std::make_shared<Clock>();
+    m_clock->addAfterCallback([this](Clock::time_t) {
+        timeUpdate();
+    });
+    m_clock->addBeforeCallback([this](Clock::time_t miliseconds) {
+        beforeTimeUpdate(miliseconds);
+    });
+    buildRpcReflection();
+}
+
+ISequencer::Status Sequencer::status() const
+{
+    return m_status;
+}
+
+mu::async::Channel<ISequencer::Status> Sequencer::statusChanged() const
+{
+    return m_statusChanged;
+}
+
+bool Sequencer::play()
+{
+    m_nextStatus = PLAYING;
+    return true;
+}
+
+void Sequencer::pause()
+{
+    m_nextStatus = STOPED;
+}
+
+void Sequencer::stop()
+{
+    m_nextStatus = STOPED;
+    rewind();
+}
+
+void Sequencer::seek(unsigned long miliseconds)
+{
+    std::lock_guard guard(m_syncMutex);
+    m_nextSeek = miliseconds;
+}
+
+void Sequencer::rewind()
+{
+    seek(0);
+}
+
+void Sequencer::initMIDITrack(ISequencer::track_id id)
+{
+    createMIDITrack(id);
+}
+
+void Sequencer::setMIDITrack(ISequencer::track_id id, const std::shared_ptr<mu::midi::MidiStream>& stream)
```

Comment:
There may be a problem, we do not know which id is free.
Someone will add a track with id 2, someone else will also add a track with id 2 - and spoil the track of the first.